ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,525 @@
+} netevent_event_notify_context_t;
+
+void
+_ebpf_netevent_push_event(_In_ netevent_event_md_t* netevent_event)
+{
+    // Logging may delay the event processing, consider enabling only for debugging or if the calling frequency for a
+    // specific use case is low.
+    // EBPF_EXT_LOG_ENTRY();
+
+    if (netevent_event == NULL) {
+        return;
+    }
+
+    ebpf_result_t result;
+    ebpf_extension_hook_client_t* client_context = NULL;
+    netevent_event_notify_context_t netevent_event_notify_context = {0};
+    uint64_t event_size = netevent_event->event_data_end - netevent_event->event_data_start;
+
+    // Currently, the verifier does not support read-only contexts, so we need to copy the event data, rather than
+    // directly passing the existing pointers.
+    // Verifier feature proposal: https://github.com/vbpf/ebpf-verifier/issues/639
+    if (event_size > _event_buffer_size) {
+        // If the event buffer is too small, attempt to resize it.
+        uint8_t* new_event_buffer =
+            (uint8_t*)ExAllocatePoolUninitialized(NonPagedPoolNx, event_size, EBPF_EXTENSION_POOL_TAG);
+        if (new_event_buffer == NULL) {
+            EBPF_EXT_LOG_MESSAGE(
+                EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                EBPF_EXT_TRACELOG_KEYWORD_NETEVENT,
+                "Failed to resize the event buffer - event lost");
+            return;
+        }
+        if (_event_buffer) {
+            ExFreePool(_event_buffer);
+        }
+        _event_buffer = new_event_buffer;
+        _event_buffer_size = event_size;
+    }
+    memcpy(_event_buffer, netevent_event->event_data_start, event_size);
+    netevent_event_notify_context.netevent_event_md.event_data_start = _event_buffer;
+    netevent_event_notify_context.netevent_event_md.event_data_end = _event_buffer + event_size;
+
+    // For each attached client call the netevent hook.
+    client_context = ebpf_extension_hook_get_next_attached_client(_ebpf_netevent_event_hook_provider_context, NULL);
+    while (client_context != NULL) {
+        NTSTATUS status = 0;
+        if (ebpf_extension_hook_client_enter_rundown(client_context)) {
+            result = ebpf_extension_hook_invoke_program(
+                client_context, &netevent_event_notify_context.netevent_event_md, (uint32_t*)&status);
+            if (result != EBPF_SUCCESS) {
+                EBPF_EXT_LOG_MESSAGE(
+                    EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                    EBPF_EXT_TRACELOG_KEYWORD_NETEVENT,
+                    "netevent_ebpf_extension_hook_invoke_program failed");
+            }
+            ebpf_extension_hook_client_leave_rundown(client_context);
+        } else {
+            EBPF_EXT_LOG_MESSAGE(
+                EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                EBPF_EXT_TRACELOG_KEYWORD_NETEVENT,
+                "netevent_ebpf_extension_hook_client_enter_rundown failed");
+        }
+        // If the client returns a non-zero value, stop calling the other clients.
```

Comment:
genuine question: why/how does an error from one client preclude us from calling the remaining ones?